consensus/beacon/oplegacy.go

```diff
@@ -0,0 +1,74 @@
+package beacon
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type OpLegacy struct{}
+
+func (o *OpLegacy) Author(header *types.Header) (common.Address, error) {
+	return header.Coinbase, nil
+}
+
+func (o *OpLegacy) VerifyHeader(chain consensus.ChainHeaderReader, header *types.Header) error {
+	// redundant check to guarantee DB consistency
+	parent := chain.GetHeader(header.ParentHash, header.Number.Uint64()-1)
+	if parent == nil {
+		return consensus.ErrUnknownAncestor
+	}
+	return nil // legacy chain is verified by block-hash reverse sync otherwise
+}
+
+func (o *OpLegacy) VerifyHeaders(chain consensus.ChainHeaderReader, headers []*types.Header) (chan<- struct{}, <-chan error) {
+	quit := make(chan struct{}, 1)
+	result := make(chan error, len(headers))
+	for range headers {
+		result <- nil
+	}
+	return quit, result
+}
+
+func (o *OpLegacy) VerifyUncles(chain consensus.ChainReader, block *types.Block) error {
+	return nil
+}
+
+func (o *OpLegacy) Prepare(chain consensus.ChainHeaderReader, header *types.Header) error {
+	return fmt.Errorf("cannot prepare for legacy block header: %s (num %d)", header.Hash(), header.Number)
+}
+
+func (o *OpLegacy) Finalize(chain consensus.ChainHeaderReader, header *types.Header, state *state.StateDB, txs []*types.Transaction, uncles []*types.Header, withdrawals []*types.Withdrawal) {
+	panic(fmt.Errorf("cannot finalize legacy block header: %s (num %d)", header.Hash(), header.Number))
```

Comment:
We are hitting this panic in a unit test.

Reply:
@trianglesphere which unit test? In the monorepo? I can help look into it

Reply:
Ah I see, `TestEthClientHistoricalBackend`. It shouldn't be calling `Finalize` there. Will investigate

Reply:
@protolambda it's getting called by CreateChain. I played around with it, but it was tricky to fix

Reply:
Yep, the legacy blocks that are supposed to go into the post-bedrock node get EVM-processed upon importing, and if we use the Legacy-consensus then it's going to try to use that legacy consensus to seal the blocks.
To get around it we need to start with regular consensus, to fill our DB with the blocks (and generated receipts etc. upon importing), and then change to legacy consensus once the legacy content is prepared. I'll push a commit with a fix